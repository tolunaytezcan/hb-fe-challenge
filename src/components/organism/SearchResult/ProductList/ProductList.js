import { useEffect } from 'react';

import ProductCard from 'components/molecules/ProductCard/ProductCard';
import Pagination from 'components/organism/Pagination/Pagination';
import { useFilter, useProducts } from 'hooks';

import styles from 'components/organism/SearchResult/ProductList/ProductList.module.scss';
import Search from 'assets/Search.svg';
import Spinner from 'assets/Spinner.svg';
import NoResult from 'assets/NoResult.svg';

const ProductList = () => {
  const { products, isLoading, isError } = useProducts();
  const { selectedFilters, setSelectedFilters } = useFilter();

  useEffect(() => {
    if (products?.length < 12) {
      setSelectedFilters(prev => ({ ...prev, page: 1 }));
    }
  }, [products]);

  const lastIndex = selectedFilters.page * 12;
  const firstIndex = lastIndex - 12;
  const results = products?.length < 12 ? products : products?.slice(firstIndex, lastIndex);

  if (isLoading) {
    return (
      <div className={styles.NoResultWrapper}>
        <div className={styles.NoResultContainer}>
          <img src={Spinner} alt="Spinner" width={72} height={72} />
          <span className={styles.NoResultText}>Yükleniyor</span>
        </div>
      </div>
    );
  }

  if (isError) {
    return (
      <div className={styles.NoResultWrapper}>
        <div className={styles.NoResultContainer}>
          <img src={NoResult} alt="No Product" width={72} height={72} />
          <span className={styles.NoResultText}>Hata! Üzgünüz bir sorunla karşılaştık.</span>
        </div>
      </div>
    );
  }

  if (results.length === 0) {
    return (
      <div className={styles.NoResultWrapper}>
        <div className={styles.NoResultContainer}>
          <img src={Search} alt="No Product" width={72} height={72} />
          <span className={styles.NoResultText}>
            {selectedFilters?.search} ile ilgili sonuç bulunamamıştır.
          </span>
        </div>
      </div>
    );
  }

  const paginate = pageNumber => {
    setSelectedFilters({ ...selectedFilters, page: pageNumber });
    window.scrollTo(0, 0);
  };

  return (
    <div>
      <div className={styles.productListContainer}>
        {results?.map(product => (
          <ProductCard key={product.id} product={product} />
        ))}
      </div>
      <div>
        <Pagination
          numberOfPages={Math.ceil(products?.length / 12)}
          paginate={paginate}
          currentPage={selectedFilters.page}
        />
      </div>
    </div>
  );
};

export default ProductList;
