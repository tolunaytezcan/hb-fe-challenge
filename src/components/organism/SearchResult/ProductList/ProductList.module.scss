@import 'styles/_variables';

.productListContainer {
  display: grid;
  grid-template-columns: repeat(4, 1fr);
  justify-content: center;
  gap: $small;

  @media (max-width: 1440px) {
    grid-template-columns: repeat(4, minmax(211px, 349.33px));
  }
  @media (max-width: 1080px) {
    grid-template-columns: repeat(3, minmax(211px, 349.33px));
  }
  @media (max-width: 768px) {
    grid-template-columns: repeat(2, minmax(211px, 349.33px));
  }
  @media (max-width: 512px) {
    grid-template-columns: 1fr;
  }
}

.NoResultWrapper {
  display: flex;
  justify-content: center;
  margin-top: 50px;
}
.NoResultContainer {
  display: flex;
  flex-direction: column;
  align-items: center;
}
.NoResultText {
  font-size: $xmedium;
  font-weight: $fw-500;
  color: $black-100;
  margin-bottom: $small;
}
