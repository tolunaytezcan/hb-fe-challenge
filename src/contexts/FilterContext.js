import { createContext, useState, useEffect, useMemo } from 'react';
import { useHistory, useLocation } from 'react-router-dom';

const FilterContext = createContext();

export default FilterContext;

function useQuery() {
  const { search } = useLocation();
  return useMemo(() => new URLSearchParams(search), [search]);
}

export const FilterProvider = ({ children }) => {
  const history = useHistory();
  const [selectedFilters, setSelectedFilters] = useState({
    color: '',
    brand: '',
    sort: '',
    search: '',
    page: 1
  });

  const clearFilters = () => {
    setSelectedFilters({
      color: '',
      sort: '',
      search: '',
      brand: '',
      page: 1
    });
  };

  let query = useQuery();

  const colorQuery = query.get('color');
  const brandQuery = query.get('brand');
  const sortQuery = query.get('sort');
  const searchQuery = query.get('search');
  const pageQuery = query.get('page');

  useEffect(() => {
    setSelectedFilters(prev => ({ ...prev, brand: brandQuery }));
    setSelectedFilters(prev => ({ ...prev, color: colorQuery }));
    setSelectedFilters(prev => ({ ...prev, sort: sortQuery }));
    setSelectedFilters(prev => ({ ...prev, search: searchQuery }));
    setSelectedFilters(prev => ({ ...prev, page: +pageQuery }));
  }, []);

  useEffect(() => {
    history.push(
      `/?brand=${selectedFilters.brand}&color=${selectedFilters.color}&sort=${selectedFilters.sort}&search=${selectedFilters.search}&page=${selectedFilters.page}`
    );
  }, [selectedFilters]);

  const values = { selectedFilters, setSelectedFilters, clearFilters };

  return <FilterContext.Provider value={values}>{children}</FilterContext.Provider>;
};
