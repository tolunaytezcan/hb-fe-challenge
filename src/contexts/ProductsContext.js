import { useState, useEffect, createContext } from 'react';
import axios from 'axios';
import { baseURL } from 'api';

const ProductsContext = createContext();

export default ProductsContext;

export const ProductsProvider = ({ children }) => {
  const [products, setProducts] = useState([]);
  const [isLoading, setIsLoading] = useState(false);
  const [isError, setIsError] = useState(false);

  useEffect(() => {
    const getProducts = async () => {
      setIsLoading(true);
      try {
        const response = await axios.get(baseURL);
        setProducts(response.data);
      } catch (error) {
        setIsError(true);
      }
      setIsLoading(false);
    };

    !products.length > 0 && getProducts();
  }, [products]);

  const values = { products, isLoading, isError };

  return <ProductsContext.Provider value={values}>{children}</ProductsContext.Provider>;
};
