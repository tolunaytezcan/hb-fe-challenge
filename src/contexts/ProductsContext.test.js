import { render } from '@testing-library/react';
import ProductList from 'components/organism/SearchResult/ProductList/ProductList';
import ProductsContext from 'contexts/ProductsContext';
import FilterContext from './FilterContext';

describe('Product context test', () => {
  const products = [
    {
      id: 1,
      brand: 'Apple',
      title: 'iPhone 11 64GB',
      price: 13950,
      color: 'Siyah',
      images: [
        {
          image: 'https://productimages.hepsiburada.net/s/49/400-592/10986385899570.jpg'
        }
      ],
      discountRate: 12,
      createdDate: '2022-07-02T17:51:41.382Z'
    }
  ];

  let selectedFilters = {
    color: '',
    brand: '',
    sort: '',
    search: '',
    page: 1
  };

  const setSelectedFilters = () => {
    selectedFilters = {
      color: '',
      brand: '',
      sort: '',
      search: '',
      page: 1
    };
  };

  it('should be render correctly', () => {
    const component = render(
      <FilterContext.Provider
        value={{
          selectedFilters,
          setSelectedFilters
        }}>
        <ProductsContext.Provider value={{ products }}>
          <ProductList />
        </ProductsContext.Provider>
      </FilterContext.Provider>
    );

    expect(component).toMatchSnapshot();
  });
});
