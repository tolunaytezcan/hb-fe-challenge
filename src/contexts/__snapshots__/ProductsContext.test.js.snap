// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product context test should be render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="productListContainer"
        />
        <div>
          <div
            class="PaginationContainer"
          >
            <span
              class="arrowButton FirstArrow"
            >
              &lt;&lt;
            </span>
            <span
              class="arrowButton"
            >
              &lt;
            </span>
            <ul
              class="PaginationList"
            >
              <li
                class="activeNumber"
              >
                <span
                  class="numberButton"
                >
                  1
                </span>
              </li>
            </ul>
            <span
              class="arrowButton"
            >
              &gt;
            </span>
            <span
              class="arrowButton LastArrow"
            >
              &gt;&gt;
            </span>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="productListContainer"
      />
      <div>
        <div
          class="PaginationContainer"
        >
          <span
            class="arrowButton FirstArrow"
          >
            &lt;&lt;
          </span>
          <span
            class="arrowButton"
          >
            &lt;
          </span>
          <ul
            class="PaginationList"
          >
            <li
              class="activeNumber"
            >
              <span
                class="numberButton"
              >
                1
              </span>
            </li>
          </ul>
          <span
            class="arrowButton"
          >
            &gt;
          </span>
          <span
            class="arrowButton LastArrow"
          >
            &gt;&gt;
          </span>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`product context should be rendered should be rendered products list 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="productListContainer"
        />
        <div>
          <div
            class="PaginationContainer"
          >
            <span
              class="arrowButton FirstArrow"
            >
              &lt;&lt;
            </span>
            <span
              class="arrowButton"
            >
              &lt;
            </span>
            <ul
              class="PaginationList"
            >
              <li
                class="activeNumber"
              >
                <span
                  class="numberButton"
                >
                  1
                </span>
              </li>
            </ul>
            <span
              class="arrowButton"
            >
              &gt;
            </span>
            <span
              class="arrowButton LastArrow"
            >
              &gt;&gt;
            </span>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="productListContainer"
      />
      <div>
        <div
          class="PaginationContainer"
        >
          <span
            class="arrowButton FirstArrow"
          >
            &lt;&lt;
          </span>
          <span
            class="arrowButton"
          >
            &lt;
          </span>
          <ul
            class="PaginationList"
          >
            <li
              class="activeNumber"
            >
              <span
                class="numberButton"
              >
                1
              </span>
            </li>
          </ul>
          <span
            class="arrowButton"
          >
            &gt;
          </span>
          <span
            class="arrowButton LastArrow"
          >
            &gt;&gt;
          </span>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
